### PR TITLE
[JENKINS-60677] Option to cancel builds when orphaning items

### PR DESCRIPTION
**Pull request by reda-alaoui**
See [JENKINS-60677](https://issues.jenkins-ci.org/browse/JENKINS-60677).

### Proposed changelog entries

* Entry 1: JENKINS-60677, Add an option allowing to cancel builds of orphaned items.

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change).
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [X] Appropriate autotests or explanation to why this change has no tests

### CODE REVIEW

**Review by jtnord — CHANGES_REQUESTED** [review]
I am concerned about waiting indefinatly for something that may never occur (a build finishing correctly when asked) along with handling things still in the queue and pipelines.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jglick) [reply src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:114]: https://www.jenkins.io/doc/developer/plugin-development/pipeline-integration/#constructor-vs-setters

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jglick) [reply src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:276]: Indeed you should not do this. Cut off the search when encountering a completed build. (Not totally correct, since it is possible for there to be a running build older than the last completed build, but good enough as a heuristic.) The iterator is lazy so you can keep the loop as is, so long as there is a break condition.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jglick) [reply src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:278]: You would need a separate code path looking for `Queue.Item`’s associated with the job and `cancel`’ing them.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:281]: Best to define a new `CauseOfInterruption`.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jglick) [reply src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:277]: I believe this will work for Pipeline builds, but indeed it would be nice to have test coverage proving it.

Would I think need a `test`-scoped dep on `workflow-multibranch` and the `tests` classifier of `scm-api` to make a multibranch project governed by a mock branch source that you can manipulate.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jglick) [reply src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:291]: I guess we are waiting because https://github.com/jenkinsci/cloudbees-folder-plugin/blob/53bf95c744c8a0ff468ef6d2b699796d2186075e/src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java#L318 relies on it? Definitely there should be a time limit waiting for an interrupt to be processed. It is not generally a big deal if we have to skip one of the orphaned projects because it has a build still running; it should just be processed the next time branch indexing runs.

File: src/main/resources/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy/config.jelly
Comment (jglick) [inline src/main/resources/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy/config.jelly:28]:
```suggestion
  <f:entry field="abortBuilds">
    <f:checkbox title="${%Abort builds}"/>
```

conventionally, but make sure to add a `help-abortBuilds.html` with details.

**Review by jglick — COMMENTED** [review]
Looking promising.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (reda-alaoui) [reply src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:276]: Ceiled with a max wait of 1 minute for all aborted builds.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:266]: FYI possibly already covered by https://github.com/jenkinsci/jenkins/blob/1a42044347502c59cd9c39d1be67f654df02908a/core/src/main/java/jenkins/model/ParameterizedJobMixIn.java#L486

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:283]: use `System.nanoTime` for comparisons

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/OrphanedParent.java
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/OrphanedParent.java:39]:
```suggestion
		this.orphanedParentFullName = orphanedParent.getFullDisplayName();
```

(if indeed used only for display)

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:105]:
```suggestion
     * @param daysToKeep     how old a branch must be to remove.
     * @param numToKeep      how many branches to keep.
```

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:286]: Why the queue rather than just iterating the list with a simple `for`-loop?

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:291]: By now I have already forgotten why we need to wait for the aborted builds to complete at all…so it would be a good idea to add a comment explaining the need for the benefit of future generations.

File: src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java
Comment (jglick) [inline src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java:322]: Or just

```suggestion
        folder.getItem("kid").setDefinition(new CpsFlowDefinition("sleep(Long.MAX_VALUE)", true));
```

should suffice I think.

And does this really work at all? You do not have a test dep on `workflow-basic-steps` nor `workflow-durable-task-step`.

If you add a test dep on the `tests` classified of `workflow-support` then you get the `semaphore` test step which is handy for things like this: can have the test wait for the build to get a well-defined point and pause before proceeding.

File: src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java
Comment (jglick) [inline src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java:844]: What is the point of the `workflow-multibranch` test dep if you are not actually creating a `MultiBranchProject`?

**Review by jglick — COMMENTED** [review]
Code changes look right to me. Tests could use some work. Sorry for the long delay in reviews, I am a bit overwhelmed, but I just remembered this PR and that it was important so I will try to help more promptly.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (reda-alaoui) [reply src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:266]: I was going to use this at first, but then I noticed `Queue#cancel(Task)` only cancel the first item of the list:
```java
    /**
     * Cancels the item in the queue. If the item is scheduled more than once, cancels the first occurrence.
     *
     * @return true if the project was indeed in the queue and was removed.
     *         false if this was no-op.
     */
    public boolean cancel(Task p) {
        lock.lock();
        try { try {
            LOGGER.log(Level.FINE, "Cancelling {0}", p);
            for (WaitingItem item : waitingList) {
                if (item.task.equals(p)) {
                    return item.cancel(this);
                }
            }
            // use bitwise-OR to make sure that both branches get evaluated all the time
            return blockedProjects.cancel(p) != null | buildables.cancel(p) != null;
        } finally { updateSnapshot(); } } finally {
            lock.unlock();
        }
    }
```
Not being acquainted to the Jenkins model, I chose the safety by cancelling the list of items.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (reda-alaoui) [reply src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:286]: Mostly for conceptual reasons. Also, polling on a queue frees memory faster.

File: src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java
Comment (reda-alaoui) [reply src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java:322]: You are correct, the build was doomed to fail but couldn't because abort was too fast.
I fixed that with the `semaphore` step.

File: src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java
Comment (reda-alaoui) [reply src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java:844]: None indeed. Dependencies cleaned.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jglick) [reply src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:266]: Probably safest to redundantly cancel all items as you do here. In practice it would rarely matter since queue items would be coalesced; they would not be when using parameterized builds, but of course branch projects are much less likely to be parameterized.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jglick) [reply src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:290]: What is the concept here? And why would we care about a few bytes of a `LinkedList`? Seems like needless complexity.

File: src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java
Comment (jglick) [inline src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java:322]: This pairs with `SemaphoreStep.waitForStart`.

File: src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java
Comment (jglick) [inline src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java:844]: Note: not as realistic as using `workflow-multibranch`, but simpler to set up and probably good enough for this purpose.

File: src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java
Comment (reda-alaoui) [reply src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java]: Done

File: src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java
Comment (jglick) [inline src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java:315]:
```suggestion
    public void runningWorkflowJobBuildWithAbortBuildsOption() throws Exception {
```

Comment (jglick) [reply src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java:315]: or
```suggestion
    public void runningPipelineBuildWithAbortBuildsOption() throws Exception {
```

File: src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java
Comment (jglick) [inline src/test/java/com/cloudbees/hudson/plugins/folder/computed/ComputedFolderTest.java:324]: I think this is the default for Pipeline anyway, so could
```suggestion
```

**Review by jglick — APPROVED** [review]
Nice! Will leave it to a plugin maintainer to merge. @jtnord who would that be?

**Comment by reda-alaoui** [comment]
Ping :)

**Review by olamy — APPROVED** [review]

**Comment by reda-alaoui** [comment]
Ping :pray:

**Comment by reda-alaoui** [comment]
Merge conflict fixed.

**Comment by reda-alaoui** [comment]
Looks like f6905c07c8e1d3ca204f193062fd1fd2f253cbf5 changed the `pom.xml` end of line from LF to CRLF in master. Was that wanted? Do I have to change the `pom.xml` line ending in this PR?

**Comment by jglick** [comment]
> Was that wanted?

No. https://issues.apache.org/jira/browse/MRELEASE-899

> Do I have to change the `pom.xml` line ending to CRLF in this PR?

No, let @jtnord fix it up.

**Comment by jtnord** [comment]
> > Was that wanted?
> 
> No. https://issues.apache.org/jira/browse/MRELEASE-899
> 
> > Do I have to change the `pom.xml` line ending to CRLF in this PR?
> 
> No, let @jtnord fix it up.

Will do, I started walking though these to merge & release, will continue tomorrow.

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jtnord) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:259]: what happens if this was not an organisation folder but a regular multibranch?

IIUC owner would be the Multibranch folder, and orphaned would be a WorkflowJob, so orphaned.getAllJobs() would be `null` or empty.

does this only work for an organisation folder when an entire repository is removed?

or is this handled somewhere else?

possibly I am missing something

**Review by jtnord — COMMENTED** [review]
@reda-alaoui could you take a look at https://github.com/jenkinsci/cloudbees-folder-plugin/pull/197/files#r809952123

this is on my list of things to do, I am just currently swamped

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java
Comment (jglick) [reply src/main/java/com/cloudbees/hudson/plugins/folder/computed/DefaultOrphanedItemStrategy.java:259]: https://github.com/jenkinsci/jenkins/blob/1a807197ba0c574c6f9c79803fcd1d5dc14fbf43/core/src/main/java/hudson/model/Job.java#L506-L509 ?

**Review by jtnord — APPROVED** [review]

**Pull request merged**
